core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -418,102 +422,120 @@ public HashDistributionTaskSource(
+                for (Split split : splitBatch.getSplits()) {
+                    int bucket = bucketNodeMap.getBucket(split);
+                    int partition = getPartitionForBucket(bucket);
 
-            finished = true;
-            return result;
+                    if (!bucketNodeMap.isDynamic()) {
+                        HostAddress requiredAddress = bucketNodeMap.getAssignedNode(split).get().getHostAndPort();
+                        Set<HostAddress> existingRequirement = partitionToNodeMap.get(partition);
+                        if (existingRequirement.isEmpty()) {
+                            existingRequirement.add(requiredAddress);
+                        }
+                        else {
+                            checkState(
+                                    existingRequirement.contains(requiredAddress),
+                                    "Unable to satisfy host requirement for partition %s. Existing requirement %s; Current split requirement: %s;",
+                                    partition,
+                                    existingRequirement,
+                                    requiredAddress);
+                            existingRequirement.removeIf(host -> !host.equals(requiredAddress));
+                        }
+                    }
+
+                    if (!split.isRemotelyAccessible()) {
+                        Set<HostAddress> requiredAddresses = ImmutableSet.copyOf(split.getAddresses());
+                        verify(!requiredAddresses.isEmpty(), "split is not remotely accessible but the list of addresses is empty: %s", split);
+                        Set<HostAddress> existingRequirement = partitionToNodeMap.get(partition);
+                        if (existingRequirement.isEmpty()) {
+                            existingRequirement.addAll(requiredAddresses);
+                        }
+                        else {
+                            Set<HostAddress> intersection = Sets.intersection(requiredAddresses, existingRequirement);
+                            checkState(
+                                    !intersection.isEmpty(),
+                                    "Unable to satisfy host requirement for partition %s. Existing requirement %s; Current split requirement: %s;",
+                                    partition,
+                                    existingRequirement,
+                                    requiredAddresses);
+                            partitionToNodeMap.replaceValues(partition, ImmutableSet.copyOf(intersection));
+                        }
+                    }
+
+                    Multimap<PlanNodeId, Split> partitionSplits = partitionToSplitsMap.computeIfAbsent(partition, (p) -> ArrayListMultimap.create());
+                    partitionSplits.put(planNodeId, split);
+                }
+
+                if (splitBatch.isLastBatch()) {
+                    completionFuture.set(null);
```

Comment:
Unfortunately with this implementation it is not possible to cancel split loading. At very minimum it should be checked if the future is cancelled before loading the next batch. But ideally it would be great to propagate cancellations to the underlying future that represents the current batch being loaded. 

A standard technique is instead of using a `SettableFuture` to extend the `AbstractFuture` and implement the `interruptTask` method.

For example you can structure it like

```
class SplitLoadingFuture extend AbstractFuture<LoadedSplits> {

private final SplitSource splitSource;
private final PlanNodeId planNodeId;
private ListenableFuture<SplitBatch> future;

// called to initiate loading and to load next batch if not finished
public synchronized load(){
   ...
   
   if(done){
     set(...)
   }
   else {
     load()
   }
}

protected synchronized interruptTask(){
  if(future != null){
      future.cancel(true);
      future = null;
  }
}

}
```